pkg/providers/cf/provider.go

```diff
@@ -0,0 +1,363 @@
+				Command:                      *proc.Command,
+				DiskQuota:                    fmt.Sprintf("%d", proc.DiskInMB),
+				HealthCheckType:              cfTypes.AppHealthCheckType(proc.HealthCheck.Type),
+				HealthCheckHTTPEndpoint:      *proc.HealthCheck.Data.Endpoint,
+				HealthCheckInvocationTimeout: uint(*proc.HealthCheck.Data.InvocationTimeout),
+				Instances:                    &procInstances,
+				LogRateLimitPerSecond:        fmt.Sprintf("%d", proc.LogRateLimitInBytesPerSecond),
+				Memory:                       fmt.Sprintf("%dMB", proc.MemoryInMB),
+				// Timeout
+				ReadinessHealthCheckType:         cfTypes.AppHealthCheckType(proc.ReadinessCheck.Type),
+				ReadinessHealthCheckHttpEndpoint: *proc.ReadinessCheck.Data.Endpoint,
+				ReadinessHealthInvocationTimeout: uint(*proc.ReadinessCheck.Data.InvocationTimeout),
+				ReadinessHealthCheckInterval:     uint(*proc.ReadinessCheck.Data.Interval),
+				Lifecycle:                        string(app.Lifecycle.Type),
+			})
+		}
+		routeOpts := client.NewRouteListOptions()
+		routeOpts.SpaceGUIDs.EqualTo(remoteSpace.GUID)
+		routes, err := cfClient.Routes.ListForAppAll(ctx, app.GUID, routeOpts)
+		if err != nil {
+			return nil, fmt.Errorf("error getting processes: %v", err)
+		}
+		appRoutes := cfTypes.AppManifestRoutes{}
+		for _, r := range routes {
+			appRoutes = append(appRoutes, cfTypes.AppManifestRoute{
+				Route:    r.URL,
+				Protocol: cfTypes.AppRouteProtocol(r.Protocol),
+				// TODO: Options: loadbalancing?
+			})
+		}
+
+		// FIXME: uncomment this when we know how to handle them
+		// allBuildpacks, err := c.cfClient.Buildpacks.ListAll(ctx, nil)
+		// if err != nil {
+		// 	return fmt.Errorf("error getting buildpacks: %v", err)
+		// }
+
+		// FIXME: uncomment this when we know how to handle them
+		// stacks, err := c.cfClient.Stacks.ListAppsOnStackAll(ctx, nil)
+		// if err != nil {
+		// 	return fmt.Errorf("error getting buildpacks: %v", err)
+		// }
+
+		// FIXME: uncomment this when we know how to handle them
+		// serviceOfferingOpts := client.NewServiceOfferingListOptions()
+		// serviceOfferingOpts.SpaceGUIDs.EqualTo(remoteSpace.GUID)
+		// serviceOfferings, err := c.cfClient.ServiceOfferings.ListAll(ctx, serviceOfferingOpts)
+		// if err != nil {
+		// 	return fmt.Errorf("error getting service offerings: %v", err)
+		// }
+
+		appManifest := cfTypes.AppManifest{
+			Name: app.Name,
+			Env:  appEnv.EnvVars, //TODO: Running, staging, appEnvVar, SystemEnvVar
+			Metadata: &cfTypes.AppMetadata{
+				Labels:      app.Metadata.Labels,
+				Annotations: app.Metadata.Annotations,
+			},
+			Processes: &appProcesses,
+			Routes:    &appRoutes,
+			// AppManifestProcess
+			// Buildpacks  //<--- There is the buildpack list but not per app
+			// RandomRoute
+			// NoRoute
+			// Services
+			// Sidecars
+			// Stack
+		}
+		if helpers.WriteToYAMLFile(appManifest, fmt.Sprintf("manifest_%s_%s.yaml", spaceName, appManifest.Name)) != nil {
```

Comment:
why do you need to write to a file here? I don't see it checking for a flag for this purpose. I would expect the call to return the list of apps and let the caller perform the write operation. This function is generating the CF manifest via REST API, it should not, IMHO, write the manifest to a file.

Reply:
We discussed this in a meeting but maybe I misunderstood the conclusion.
What I got is:
- Local discovery: the library generates the discovery manifest, return the object and let the invoker decide what to do.
- Live discovery: if the output path is set, the library generate the discovery manifest,  writes the files (one for each app) and return an empty list. If the output path is unset the library return the list of objects.

The problem is related to the quantity of the generated data.

However, returning the list of objects no matter what is even easier, so i can change it. Let me know what you prefer